### ISSUE TITLE
integer constant overflow on gccgo and 32-bit

### ISSUE THREAD

**Issue by zachjs** [go-mgo/mgo#310]
After upgrading to the latest version on v2, I'm running into this error when building with gccgo.
```
[2016/08/02 19:13:01.577] # gopkg.in/mgo.v2/bson
[2016/08/02 19:13:01.577] vendor/src/gopkg.in/mgo.v2/bson/json.go:320:11: error: integer constant overflow
[2016/08/02 19:13:01.577]   if n <= 1<<53 {
```

Perhaps `int64` should be used explicitly here instead of `int`.

*Update:* This issue is also occurring using the standard go compiler on 32-bit Windows.
```
[2016/08/02 21:21:22.468] # gopkg.in/mgo.v2/bson
[2016/08/02 21:21:22.468] .gopath\src\gopkg.in\mgo.v2\bson\json.go:320: constant 9007199254740992 overflows int
```

**Comment by aprice** [comment]
I have the same issue on 64-bit linux when cross-compiling for 386 arches.

**Comment by pfzero** [comment]
I get the same error on Raspberry 3 Model B - armv7: `constant 9007199254740992 overflows int`.

**Comment by pfzero** [comment]
@zachjs @niemeyer I've modified from `int` to `int64` and it works fine when compiling. But I don't know if there are any other suble implications with this change, or wether there are other changes that'll produce runtime errors because of the arch difference.

**Comment by zachjs** [comment]
@pfzero Check out #311. Simply changing `int` to `int64` would cause a panic as any interface that is passed to that function will always be an `int`.

**Comment by pfzero** [comment]
@zachjs Strange because I have modified to `int64` and it works ok for me. But tbh I haven't tested it on 64-bit arch. Also, I guess that asserting an int value within int64 should be ok since int64 can assimilate int. Either way, I like your approach in #311. 

Cheers!

**Comment by halimess** [comment]
When I try to get mgo from golang version 1.6.2, it's ok.
But, in 1.6.3 version this overflow error occur during setting up

This info may be helpful to you guys, I think

**Comment by lonelycode** [comment]
This issue also exists in go 1.7:

```
gox -osarch="linux/386"
Number of parallel builds: 1

-->       linux/386: github.com/xxxx/xxxx

1 errors occurred:
--> linux/386 error: exit status 2
Stderr: # gopkg.in/mgo.v2/bson
../go/src/gopkg.in/mgo.v2/bson/json.go:320: constant 9007199254740992 overflows int
```

**Comment by niemeyer** [comment]
Fixed with @zachjs's #311. Thanks!
